#231 fix - Store users location and address searches

eladkarakuli opened the pull request:
well, the app just didn't worked for me - I tried to create events base routing (to capture when user changed location) and nothing worked.
not a single route except the "" (empty) one. So I changed the routing mechanism, so now its routing appropriate.

By activating the Routing, some unintended callings was starting to happen, I think from all the bindings and callings to updateUrl().

The server request handling was just wrong, and uninformative, I mean there is couple of functions handling requests for marker that just never get executed. say POST/PUT we're never creating new markers from an http request, as far as I know.

<!---
@huboard:{"order":229.5,"milestone_order":243,"custom_state":""}
-->

danielhers commented on main.py:
```diff
@@ -112,22 +114,53 @@ def markers(methods=["GET", "POST"]):
         marker.put()
         return make_response(json.dumps(marker.serialize(self.user)))
 
+#TODO: if not in future use - remove
```
You're right, we currently don't support insertion and deletion of markers, so it can be removed.

eladkarakuli replied:
Will be removed.

danielhers commented on main.py:
```diff
@@ -27,6 +27,8 @@
 
 
 assets_env = AssetsEnvironment('./static/', '/static')
+assets_env.debug = True
```
Should be disabled?

danielhers commented on main.py:
```diff
@@ -112,22 +114,53 @@ def markers(methods=["GET", "POST"]):
 def marker(self, key_name):
+    return(make_response(get_put_delete(self, key_name, Marker)))
+
+@app.route("/highlightpoints/", methods=['POST'])
```
Why "highlight"? I think the intention is to record the user location.

eladkarakuli replied:
we are separating the user from location, as we don't want to save user and it's location, only the location is important.

danielhers commented on main.py:
```diff
@@ -112,22 +114,53 @@ def markers(methods=["GET", "POST"]):
+
+# Post Handler For REST API
+# cls: ORM Model, implement parse and serialize functions.
+def postHandler(cls, data):
```
Use "post_handler" as per Python naming convention.

danielhers commented on models.py:
```diff
@@ -55,6 +56,38 @@ def format_description(field, value):
+class HighlightPoint(Point, Base):
+    __tablename__ = "highlight_markers"
+    __table_args__ = (
+        Index('highlight_long_lat_idx', 'latitude', 'longitude'),
```
Why do we need an index on this table?

eladkarakuli replied:
I was thinking that in the future, when we decide to make some scene from all the locations, to use an index table so particular querying will be faster, actually I'm not sure how we want to query the highlights table in the future, **so should I remove it?**
P.S I'm working on creating a R*Tree module, for better location querying, for the accidents bounding queries, It should be usable here to.

danielhers replied:
I guess it won't hurt to leave it.

danielhers commented on static/js/app.js:
```diff
@@ -165,16 +167,23 @@ $(function() {
-        navigate: function(id, start, end, showFatal, showSevere, showLight, showInaccurate, zoom, lat, lon) {
-            app.model.set("currentMarker", parseInt(id));
+        navigateNoMarker: function(start, end, showFatal, showSevere, showLight, showInaccurate, zoom, lat, lon) {
+            this.navigateSelectedMarker("" ,start, end, showFatal, showSevere, showLight, showInaccurate, zoom, lat, lon);
```
Should use `null` for empty value

danielhers commented on static/js/app.js:
```diff
@@ -165,16 +167,23 @@ $(function() {
     var AppRouter = Backbone.Router.extend({
         routes: {
             "" : "navigateEmpty",
-            "/?marker=:id&start_date=:start&end_date=:end&show_fatal=:showFatal&show_severe=:showSevere&show_light=:showLight&show_inaccurate=:showInaccurate&zoom=:zoom&lat=:lat&lon=:lon" : "navigate"
```
It was "?/" because that was the root route. Did it not work when you changed it to a different one?

eladkarakuli replied:
Yes, the routing didn't work with standard URLs for some reason, so I switched to work with the hash  '#' URLs.

danielhers replied:
Makes sense, actually, as this is what they say [in the tutorial](http://backbonetutorials.com/what-is-a-router/)

danielhers commented:
I think the intention was to record the user's actual location, as is calculated in `goToMyLocation`, and not the location that the user navigated to. So you don't really need to change the routing in the JS, I think.

Also, please write an informative commit message. Preferably it should be one commit, not two.

galraij commented:
Both his actual and also what he searched for. If a user allow to find his place it means he finds this place interesting, if he searches for a place its the same.

I think Elad said he can mention the source in the file, so we can also decide one day to ignore one of the types.



‫ב-7 במאי 2015, בשעה 10:33, ‏‏Daniel Hershcovich ‏<notifications@github.com> כתב/ה:‬

> I think the intention was to record the user's actual location, as is calculated in goToMyLocation, and not the location that the user navigated to. So you don't really need to change the routing in the JS, I think.
> 
> Also, please write an informative commit message. Preferably it should be one commit, not two.
> 
> —
> Reply to this email directly or view it on GitHub.
>

danielhers commented:
Right, both searches and the user's actual physical location, but not any location the user navigates to in the app, right?

galraij commented:
True, good point.



‫ב-7 במאי 2015, בשעה 10:41, ‏‏Daniel Hershcovich ‏<notifications@github.com> כתב/ה:‬

> Right, both searches and the user's actual physical location, but not any location the user navigates to in the app, right?
> 
> —
> Reply to this email directly or view it on GitHub.
>

eladkarakuli commented:
Thanks @danielhers for pointing out goToMyLocation, for some reason I missed the location initialization and update function. I'll change the binding so It'll work with the function. 
You can see that I bind the new location update function as "once" so It's only create 1 new highlight point from the user location, just saying :)

@galraij, yes, it'll save the locations with USER GPS/SEARCH indication.

danielhers commented on static/js/app.js:
```diff
@@ -692,11 +700,21 @@ $(function() {
+        createHighlightPoint : function(lat, lng, highlightPointType) {
+            if (isNaN(lat) || isNaN(lng) || isNaN(highlightPointType)) return;
+
+            Backbone.$.ajax({
```
Use the `$.post` shortcut, see [here](https://github.com/hasadna/anyway/blob/master/static/js/app.js#L674).

danielhers commented:
Please just update (rebase) your branch from upstream/master so we can merge it.

eladkarakuli commented:
Hi @danielhers, I fixed your note about the ajax call.
Did some refactoring at sever side, now when calling a post handler, it'll catch the error/exception if accrue and will not bounce it to the client, only debug logging will tell about an error, that is why I think we should have a **logger** of some sort, what do you think?
Also, changed the marker (now Point class) id column type to **Integer** as for Sqlite doesn't allow BIGINT used as a primary key with autoincrement, there is a [workaround](http://stackoverflow.com/questions/18835740/does-bigint-auto-increment-work-for-sqlalchemy-with-sqlite), but do we really need it?

danielhers commented on main.py:
```diff
@@ -133,14 +134,58 @@ def discussion():
+    try:
+        return request.get_json(force=True)
+    except Exception as e:
+        logging.debug("Could not get json from the request. cls:{0}. Error:{1} {2}".format(request, e.type, e.message))
```
`cls` should be `request` here

eladkarakuli replied:
sorry?

danielhers commented on main.py:
```diff
@@ -133,14 +134,58 @@ def discussion():
+# cls: ORM Model, implement parse and serialize functions.
+def parse_json_request(cls, request):
+    try:
+        data = get_json(request)
```
If `get_json` had an exception, you'll get `None` here. Better just raise the exception further or check for `None`

danielhers commented on main.py:
```diff
@@ -133,14 +134,58 @@ def discussion():
+    # at this point the marker id is just a running number, and the
+    # LMS is in the description and needs to be promoted to a DB
+    # field so we can query it. We also need to add a provider id.
+    context = {'minimal_zoom': MINIMAL_ZOOM, 'url': request.url_root}
```
Why `url_root`? I changed it to `base_url`

eladkarakuli replied:
:hankey: merge. thank you for pointing it out.

danielhers commented on models.py:
```diff
@@ -84,6 +83,38 @@ def format_description(field, value):
+    HIGHLIGHT_TYPE_USER_SEARCH = 1
+    HIGHLIGHT_TYPE_USER_GPS = 2
+
+    created = Column(DateTime, default=datetime.datetime.now)
```
If these are common to all Points, why not define them there?

eladkarakuli replied:
Do you mean the `created` column? Now its  indexed for markers and not indexed for highlights, do we want to change it?

danielhers replied:
You can define the index per class (`__table_args__`), it doesn't have to be in the column definition.

danielhers commented:
Regarding a logger, yes, I think it would be a good idea to keep error logs and notify us somehow when they occur, because an error while processing the highlight points POST requests won't be visible to anyone unless we happen to run `heroku logs`.

danielhers commented:
Regarding the primary key, `Integer` limits us to 2 billion markers. But I suppose that's fine.
Did you check the `autoincrement` works? It would mean that we don't need to give an `id` explicitly in `process.py` or in `DiscussionMarker.parse`.

eladkarakuli commented:
The highlight insertion with auto-increment works fine for me, should be the same for other objects.

eladkarakuli commented:
Does `heroku logs` enables you to see the `logging.debug` outputs?

danielhers commented:
Yes, it does.

danielhers commented on static/js/app.js:
```diff
@@ -226,7 +234,7 @@ $(function() {
         },
         fetchMarkers : function() {
             if (!this.isReady) return;
-            this.updateUrl();
```
Is the URL still getting updated when moving and zooming, after you removed this line?

eladkarakuli replied:
when moving and changing zoom, map fire the idle event, than it's re-fetching the markers, which on success trigger the reloadSidebar that calling the chooseMarker, then when there is no marker selected, it'll call updateUrl.
The udpateUrl does not fire when there is a marker selected and we do map manipulation (without closing the marker info window) do you think it's a problem?

danielhers replied:
It's probably fine, as the URL should keep pointing to the marker ID as long as the info window is open.

eladkarakuli replied:
It did reduce the app's loading time. I can't tell exactly by how much, but it feels visually better.

danielhers commented:
Well done on writing a test, too - starting to do that is something we've been planning for a while.

Do you think we'll have to reload all the accidents after merging, because of the column type change (`id`)?

eladkarakuli commented:
To apply the auto increment we have to recreate the tables with the new integer column in the DB.

danielhers commented:
Right, so we'll wait until the website is less busy before merging this and reloading the data.

danielhers commented:
Great. Will merge together with https://github.com/hasadna/anyway/pull/254 when that one is ready, and then reload the data.
Another tiny request: please use a more informative commit message, describing what changes you did to solve the issue and mentioning that you added tests.

danielhers commented:
@eladkarakuli I think it might actually not even require reloading the data. See this staging environment that works with our production DB: https://anyway-pr-243.herokuapp.com/

omerxx commented on tests/test_main.py:
```diff
@@ -0,0 +1,40 @@
+  def test_class_without_parse_method(self):
+    self.assertEqual(parse_data(self.NoParsedObject, self.marker_dummy), None)
+
+  def test_pares_marker(self):
```
shouldn't be `parse`?

Thanks for writing this test I've learnt a lot :)

eladkarakuli replied:
Good eye @omerxx, I've fix that misspelling, thanks.

I've used the standard python unit testing framework, unittest. If you like there is more info in the [docs](https://docs.python.org/2/library/unittest.html).

eladkarakuli commented:
@danielhers, no markers are showing in the link you provided.
I madded the commit message a bit more informative, also I did inform in my commit that I added tests, are you sure you seeing the full commit message by pressing the "..." button?
